cloudevents/working-drafts/minor-versions.md

```diff
@@ -0,0 +1,337 @@
+then any attempt to use the "special" access should fail when accessing an attribute within
+an event with a version that doesn't have that attribute as a core attribute. Example:
+
+```csharp
+// Create a v1.0 event
+var evt = new CloudEvent(CloudEventsSpecVersion.V1_0);
+
+// Try to access the new Xyz property, referring to the optional attribute introduced in
+// v1.1: this should fail.
+evt.Xyz = 1;
+
+// Access the xyz attribute by name, as an extension attribute.
+// This should work, even though the value wouldn't be valid in a v1.1 event.
+evt["xyz"] = "test";
+
+// Likewise accessing the property for an event parsed from elsewhere:
+evt = httpRequest.ParseCloudEvent();
+// If evt is v1.0, this should fail *even if it has an xyz attribute value that's valid in v1.1.
+int xyz = evt.Xyz;
+```
+
+Setting a value within an event where the version indicates that the attribute *is* known should
+validate the attribute value according to that version. Example:
+
+```csharp
+// Create a v1.1 event
+var evt = new CloudEvent(CloudEventsSpecVersion.V1_1);
+
+// This should work: 1 is a valid value, and xyz is known in v1.1
+evt.Xyz = 1;
+
+// This should fail: 123 is not a valid value for xyz in v1.1
+evt.Xyz = 123;
+
+// This should fail to compile, where that's feasible: xyz should always be an integer
+evt.Xyz = "test";
+
+// This should succeed at execution time
+evt["xyz"] = 123;
+
+// This should fail at execution time due to validation
+evt["xyz"] = "test";
+
+// This should fail at execution time due to validation
+evt["xyz"] = 123;
+```
+
+### "Changing" event versions
+
+An SDK may provide the ability to change the specversion of an event,
+or (preferred) create a new object based on the old one. When "upgrading" an event's specversion,
+the SDK should perform validation as if parsing the event from scratch. Examples:
+
+```csharp
+// Successful upgrade: no xyz value
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0) { Id = "..." };
+var evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful upgrade: valid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = 10
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Failed upgrade: invalid value for xyz
+var evt10 = new CloudEvent(CloudEventsSpecVersion.V1_0)
+{
+    Id = "...",
+    ["xyz"] = "test"
+};
+evt11 = evt10.WithSpecVersion(CloudEventsSpecVersion.V1_1);
+
+// Successful downgrade
+var evt11 = new CloudEvent(CloudEventsSpecVersion.V1_1)
+{
+    Id = "...",
+    Xyz = 10
+};
+evt10 = evt11.WithSpecVersion(CloudEventsSpecVersion.V1_0);
+// Access to evt10.Xyz would fail as described above;
+// access to evt10["xyz"] would return 10.
+// Setting evt10["xyz"] to "test" or 123 could fail or succeed - it's
+// up to the SDK to decide (and document!) whether downgrading retains constraints
+// for attributes that were present in the original event.
```

Comment:
I'd think this should always work since for a v1.0 event it can be either value. If you allow for the 1.1 constraints to live on, then your `evt10.Xyz` example should work if you want to be consistent.

Reply:
That ends up being slightly tricky for the C# SDK, as an event can be created with known extension attributes which can have constraints. So it would be feasible to create a 1.0 event with the xyz attribute with the same constraints as in 1.1. I *think* I'd expect that to be the behavior when downgrading. But this is definitely a detailed area we should only get into if we actually go ahead with the whole thing.